client/services/ReinforcementGeneratorService.js

```diff
@@ -0,0 +1,120 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for generating positive reinforcement when a user
+ * submits code based on the tasks that the code passes/fails on.
+ */
+
+tie.factory('ReinforcementGeneratorService', [
+  'ReinforcementObjectFactory', 'TranscriptService',
+  function(ReinforcementObjectFactory, TranscriptService) {
+
+    var getTestsKeyedByTagAndIndexes = function(tests) {
+      var splitTests = {};
+      var inputToOriginalIndex = {};
+      for (var i = 0; i < tests.length; ++i) {
+        var test = tests[i];
+        inputToOriginalIndex[test.getStringifiedInput()] = i;
+        var testTag = test.getTag();
+        if (splitTests.hasOwnProperty(testTag)) {
+          splitTests[testTag].push(test);
+        } else {
+          splitTests[testTag] = [test];
+        }
+      }
+      return {
+        splitTests: splitTests,
+        testIndexes: inputToOriginalIndex
+      };
+    };
+
+    return {
+      getReinforcement: function(task, codeEvalResult) {
+
+        var previousSnapshot = TranscriptService.getMostRecentSnapshot();
+        var previousReinforcement = null;
+        if (previousSnapshot !== null) {
+          previousReinforcement = previousSnapshot.getReinforcement();
+        }
+
+        var reinforcement = ReinforcementObjectFactory.create(task);
+
+        // Copy the previous reinforcement object if we are on the same task.
+        if (previousReinforcement !== null &&
+            task === previousReinforcement.getTask()) {
+          var previousReinforcementPassedTags =
+            previousReinforcement.getPassedTags();
+          for (var tag in previousReinforcementPassedTags) {
+            reinforcement.addToPassedTags(tag,
+              previousReinforcementPassedTags[tag]);
+          }
+          var previousReinforcementPastFailedCases =
+            previousReinforcement.getPastFailedCases();
+          for (var pastCase in previousReinforcementPastFailedCases) {
+            reinforcement.addToPastFailedCases(pastCase,
+              previousReinforcementPastFailedCases[pastCase]);
+          }
+        }
+
+        // Go through correctness tests to update reinforcement data.
+        var correctnessTestsByTagAndIndexes = getTestsKeyedByTagAndIndexes(
+          task.getCorrectnessTests());
+        var correctnessTestsByTag = correctnessTestsByTagAndIndexes.splitTests;
+        var testIndexes = correctnessTestsByTagAndIndexes.testIndexes;
+        var observedOutputs = codeEvalResult.getCorrectnessTestResults();
+        // Get results only for the last task.
+        observedOutputs = observedOutputs[observedOutputs.length - 1];
```

Comment:
Don't redefine a variable. Instead, create a new method along the lines of codeEvalResult.getLastTaskResults()